### PR TITLE
build: add a cmake target for all used linters

### PR DESCRIPTION
**Pull request by dundargoc**
build: add a cmake target for all used linters

Part 1 of the cmake saga (this one)
Part 2 is going to be checking the uncrustify version is OK, and warn if it isn't.
Part 3 is about creating a `make format` target that format everything.

### CODE REVIEW

**Comment by dundargoc** [comment]
Would like this merged before working on uncrustify CI as that will have at least 3 more optional commands if my calculations are correct.

**Comment by justinmk** [comment]
> All other program checks are done in Makefile except for lualint, which
> is checked in CMakeLists.txt instead. Having the check in the Makefile
> is more consistent

It should probably be the reverse. CMake is our source of truth. Makefile is a thin convenience, it's not supposed to become the source of truth. People should be able to get all build tasks cmake only.

Makefile tasks are not available on Windows, xcode, ninja, etc.

**Comment by dundargoc** [comment]
> CMake is our source of truth.

We live in the darkest timeline.

I'll move the function to cmake instead :+1:

File: CMakeLists.txt [hunk 1]
```diff
@@ -748,15 +745,51 @@ if(BUSTED_LUA_PRG)
...
-if(LUACHECK_PRG)
-  add_custom_target(lualint
-    COMMAND ${LUACHECK_PRG} -q runtime/ scripts/ src/ test/
-    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
-else()
-  add_custom_target(lualint false
-    COMMENT "lualint: LUACHECK_PRG not defined")
-endif()
-
+find_program(LUACHECK luacheck)
```
Comment (justinmk) [inline CMakeLists.txt:748]: the `_PRG` suffix convention was intentional and used throughout our cmake files. Why revert that?

File: CMakeLists.txt [hunk 2]
```diff
@@ -748,15 +745,51 @@ if(BUSTED_LUA_PRG)
...
+
+set(TARGET "lualint")
+add_custom_target(${TARGET}
+  COMMAND ${CMAKE_COMMAND}
+  -DPROGRAM=${LUACHECK}
+  -DPROJECT_ROOT=${PROJECT_SOURCE_DIR}
+  -DTARGET=${TARGET}
+  -P ${PROJECT_SOURCE_DIR}/cmake/lint.cmake)
+
+set(TARGET "stylua")
```
Comment (justinmk) [inline CMakeLists.txt:762]: all lint tasks should have `lint` suffix. Or better, a `lint` prefix, since we're changing things already.

File: Makefile [hunk 3]
```diff
@@ -47,13 +47,7 @@ endif
...
 ifeq (,$(BUILD_TOOL))
   ifeq (Ninja,$(CMAKE_GENERATOR))
-    ifneq ($(shell $(CMAKE_PRG) --help 2>/dev/null | grep Ninja),)
-      BUILD_TOOL = ninja
-    else
-      # User's version of CMake doesn't support Ninja
-      BUILD_TOOL = $(MAKE)
-      CMAKE_GENERATOR := Unix Makefiles
-    endif
+    BUILD_TOOL = ninja
```
Comment (justinmk) [inline Makefile:50]: why remove this fallback?

File: cmake/lint.cmake [hunk 4]
```diff
@@ -0,0 +1,35 @@
+if(PROGRAM)
```
Comment (justinmk) [inline cmake/lint.cmake:1]: If we're trying to de-clutter the main `CMakeLists.txt`, can we move all of the `add_custom_target` calls into this file?  It doesn't need to be parameterized.

File: CMakeLists.txt (hunk 1 above)
Comment (dundargoc) [reply CMakeLists.txt:748]: Found it too wordy, don't mind either way.

File: Makefile (hunk 3 above)
Comment (dundargoc) [reply Makefile:50]: It's dead code. It's only useful if user's version of cmake doesn't support ninja, but our minimum required cmake version is 3.10, which supports ninja.

File: cmake/lint.cmake (hunk 4 above)
Comment (dundargoc) [reply cmake/lint.cmake:1]: I'd need to create an additional file in that case due to the limitations of `add_custom_target`:
- it can't silence output
- overall very inflexible since it won't allow you to run other commands.

File: CMakeLists.txt (hunk 1 above)
Comment (justinmk) [reply CMakeLists.txt]: `_PRG` convention should be kept

File: Makefile (hunk 3 above)
Comment (justinmk) [reply Makefile:50]: This fallback was for users that don't have `ninja` installed.

File: cmake/lint.cmake (hunk 4 above)
Comment (justinmk) [reply cmake/lint.cmake:1]: ok but this file isn't doing much then. It's just shuffling some text into a switch-case.

File: ci/run_lint.sh [hunk 5]
```diff
@@ -12,7 +12,7 @@ rm -f "$END_MARKER"
 
 # Run all tests if no input argument is given
 if (($# == 0)); then
-  tests=('clint-full' 'lualint' 'pylint' 'shlint' 'check-single-includes')
+  tests=('lint_c_full' 'lint_lua' 'lint_py' 'lint_sh' 'check-single-includes')
```
Comment (justinmk) [inline ci/run_lint.sh:15]: I don't think underscores are used in any of our other build targets, so let's stay consistent. Could use hyphens I guess, which is in some of our build targets, but that is annoying to type/complete.

```suggestion
  tests=('lintcfull' 'lintlua' 'lintpy' 'lintsh' 'check-single-includes')
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -748,15 +745,51 @@ if(BUSTED_LUA_PRG)
...
+find_program(FLAKE8_PRG flake8)
+find_program(UNCRUSTIFY_PRG uncrustify)
+
+set(TARGET "lint_lua")
+add_custom_target(${TARGET}
+  COMMAND ${CMAKE_COMMAND}
+  -DPROGRAM=${LUACHECK_PRG}
+  -DPROJECT_ROOT=${PROJECT_SOURCE_DIR}
+  -DTARGET=${TARGET}
+  -P ${PROJECT_SOURCE_DIR}/cmake/lint.cmake)
```
Comment (justinmk) [inline CMakeLists.txt:760]: cmake is forcing us to be verbose here, nothing we can do about that. But this means `cmake/lint.cmake` gains little. Just put the extra 1-2 lines here, the indirection of `cmake/lint.cmake` isn't justified to save 1-2 lines here.

File: Makefile (hunk 3 above)
Comment (dundargoc) [reply Makefile:50]: I built neovim without ninja installed, it defaults to make (which I forgot how slow it was, dear lord).

File: cmake/lint.cmake (hunk 4 above)
Comment (dundargoc) [reply cmake/lint.cmake:1]: https://github.com/uncrustify/uncrustify/pull/3697

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -107,11 +107,11 @@ jobs:
...
 
       - if: "!cancelled()"
-        name: clint-full
-        run: ./ci/run_lint.sh clint-full
+        name: lintcfull
+        run: ./ci/run_lint.sh lintcfull
 
       - if: "!cancelled()"
-        name: stylua
+        name: lintstylua
```
Comment (justinmk) [inline .github/workflows/ci.yml:114]: why isn't this part of `lintlua` ?

File: cmake/lint.cmake [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+  if(${TARGET} STREQUAL "lintuncrustify")
```
Comment (justinmk) [inline cmake/lint.cmake:3]: why isn't this part of `lintc` ? we don't need separate build targets depending on the tool. We need "lint" tasks, not "lint with a specific tool". The tool is an implementation detail.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (dundargoc) [reply .github/workflows/ci.yml:114]: Wait, this becomes super difficult without having an external script, because we'd need four separate add_custom_target for all cases if stylua/luacheck exists/doesn't exist.

**Review by justinmk — APPROVED** [review]
This looks like a nice improvement! 

As a potential followup or bonus, it might be a good idea for our CI to use these targets, that ensures they are "alive" and working. (Except for CI jobs where we don't want to build `nvim`.)

File: cmake/lint.cmake [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      message(FATAL_ERROR "FAILED: ${TARGET}")
+    endif()
+  else()
+    message(STATUS "${TARGET}: ${LINT_PROGRAM} not found. SKIP.")
+  endif()
+endfunction()
+
+if(${TARGET} STREQUAL "lintuncrustify")
+  file(GLOB_RECURSE FILES ${PROJECT_ROOT}/src/nvim/*.[c,h])
+  lint(PROGRAM uncrustify FLAGS -c src/uncrustify.cfg -q --check FILES ${FILES} QUIET)
```
Comment (justinmk) [inline cmake/lint.cmake:26]: nit:  consistent casing is helpful, even if it's optional in cmake. The function is defined as `LINT`?

File: cmake/lint.cmake [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+  if(LINT_QUIET)
+    set(OUTPUT_QUIET OUTPUT_QUIET)
+  elseif()
+    set(OUTPUT_QUIET "")
+  endif()
+
+  find_program(PROGRAM_EXISTS ${LINT_PROGRAM})
+  if(PROGRAM_EXISTS)
+    execute_process(COMMAND ${LINT_PROGRAM} ${LINT_FLAGS} ${LINT_FILES}
```
Comment (justinmk) [inline cmake/lint.cmake:12]: Does this always run, even if there are no changes since the last run? If so, that's a regression. We can think about that in a followup PR though.

File: CMakeLists.txt [hunk 11]
```diff
@@ -748,15 +745,13 @@ if(BUSTED_LUA_PRG)
     set_target_properties(functionaltest-lua PROPERTIES FOLDER test)
 endif()
 
-if(LUACHECK_PRG)
-  add_custom_target(lualint
-    COMMAND ${LUACHECK_PRG} -q runtime/ scripts/ src/ test/
```
Comment (justinmk) [inline CMakeLists.txt:753]: The old target used `COMMAND ${LUACHECK_PRG}` and the new target uses `execute_process()`. I wonder if we still get the benefits of cmake's "dirty workspace detection".

**Comment by dundargoc** [comment]
> As a potential followup or bonus, it might be a good idea for our CI to use these targets, that ensures they are "alive" and working.

They are, albeit in a pretty roundabout way. CI lint steps runs `ci/run_lint.sh` which in turn calls `make` which in turn calls `cmake`. Turtles all the way down :)

**Comment by justinmk** [comment]
Eek, this doesn't find `luacheck` on my workspace now:

    -- lintlua: luacheck not found. SKIP.

Does it need to look in .deps ?
